src/main/java/net/minecraftforge/fluids/BlockFluidBase.java

```diff
@@ -462,36 +469,63 @@ public IBlockState getExtendedState(@Nonnull IBlockState oldState, @Nonnull IBlo
         float[][] height = new float[3][3];
         float[][] corner = new float[2][2];
         height[1][1] = getFluidHeightForRender(worldIn, pos);
-        if(height[1][1] == 1)
+        if (height[1][1] == 1)
         {
-            for(int i = 0; i < 2; i++)
+            for (int i = 0; i < 2; i++)
             {
-                for(int j = 0; j < 2; j++)
+                for (int j = 0; j < 2; j++)
                 {
                     corner[i][j] = 1;
                 }
             }
         }
         else
         {
-            for(int i = 0; i < 3; i++)
+            for (int i = 0; i < 3; i++)
             {
-                for(int j = 0; j < 3; j++)
+                for (int j = 0; j < 3; j++)
                 {
-                    if(i != 1 || j != 1)
+                    if (i != 1 || j != 1)
                     {
                         height[i][j] = getFluidHeightForRender(worldIn, pos.add(i - 1, 0, j - 1));
                     }
                 }
             }
-            for(int i = 0; i < 2; i++)
+            for (int i = 0; i < 2; i++)
             {
-                for(int j = 0; j < 2; j++)
+                for (int j = 0; j < 2; j++)
                 {
                     corner[i][j] = getFluidHeightAverage(height[i][j], height[i][j + 1], height[i + 1][j], height[i + 1][j + 1]);
                 }
             }
         }
+        
+        //check for downflow above corners
+        boolean n =  isCoveredWithFluid(worldIn, pos.add(-1, 0,  0));
+        boolean s =  isCoveredWithFluid(worldIn, pos.add( 1, 0,  0));
+        boolean w =  isCoveredWithFluid(worldIn, pos.add( 0, 0, -1));
+        boolean e =  isCoveredWithFluid(worldIn, pos.add( 0, 0,  1));
+        boolean nw = isCoveredWithFluid(worldIn, pos.add(-1, 0, -1));
+        boolean ne = isCoveredWithFluid(worldIn, pos.add(-1, 0,  1));
+        boolean sw = isCoveredWithFluid(worldIn, pos.add( 1, 0, -1));
+        boolean se = isCoveredWithFluid(worldIn, pos.add( 1, 0,  1));
```

Comment:
For rendering purposes, looking up blockstates in the world is pretty slow, so this strategy seems inefficient.
Is there any way to leverage the lookups in `getFluidHeightForRender` to avoid looking up a particular position twice?
How many `world.getBlockState` lookups do vanilla fluids do?